core/node/zkos_prover_input_generator/src/zkos_proof_data_server.rs

```diff
@@ -149,37 +198,74 @@ async fn submit_snark_proof(
     };
 
     let block_number = L2BlockNumber(payload.block_number);
-    info!("Submitting SNARK proof for block {}", block_number);
-    match conn.zkos_prover_dal()
+    info!("Received SNARK proof submission for block {}", block_number);
+    debug!("SNARK proof size: {} bytes", proof_bytes.len());
+
+    // Upload to S3 if object store is configured
+    if let Some(object_store) = &state.object_store {
+        let s3_key = format!("snark_proof_{}", block_number.0);
+        info!("Uploading SNARK proof to S3 with key: {}", s3_key);
+
+        match object_store
+            .put_raw(Bucket::ZkOSProofs, &s3_key, proof_bytes.clone())
+            .await
+        {
+            Ok(_) => {
+                info!(
+                    "Successfully uploaded SNARK proof to S3 for block {}",
+                    block_number
+                );
+            }
+            Err(err) => {
+                error!(
+                    "Failed to upload SNARK proof to S3 for block {}: {}",
+                    block_number, err
+                );
+                // Return error to client and don't save to database
+                return (
+                    StatusCode::SERVICE_UNAVAILABLE,
+                    format!("Failed to store proof: {}", err),
+                )
+                    .into_response();
+            }
+        }
+    } else {
+        debug!("Object store not configured, skipping S3 upload for SNARK proof");
+    }
+
+    // Save to database only after successful S3 upload
+    info!("Saving SNARK proof to database for block {}", block_number);
+    match conn
```

Comment:
maybe extract this in a function and reuse for both cases? Should be straightforward